Adding rack-aware partition allocation algorithm

cgtz opened the pull request:
- Added implementations of rack-aware partition allocation algorithms to ClusterMapManager

- Added associated test code for rack-aware allocation

- Added CLI with rack-aware options to PartitionManager

Previously, I had two rack-aware allocator implementations for testing/evaluation purposes.  One used the random choice strategy described here (http://marios.io/2011/08/31/multiple_choice_paradigm/), and the other (currently unused) first shuffled the nodes and then tried the node with the most free space.  The former performed better in testing, so I removed the latter from the PR.

*Time to review:* 30-40 Min.
*Primary Reviewers*: Gopal, Priyesh
*Coverage*: 100% branch/line coverage for newly added lines in ClusterMapManager (`allocatePartitions`, `allocateDisksForPartitionWithPotentialRetry`, `allocateDisksForPartition`, `getRandomDiskCandidateSample`).  No unit tests for PartitionManager and other CLI tools in ambry-tools

| Class | Class, % | Method, % | Line, % |
| --- | --- | --- | --- |
| ClusterMapManager | 66.7% (2/ 3) | 80.6% (29/ 36) | 71.6% (156/ 218) |

*Truncated generated rack aware partition layout:* https://gist.github.com/cgtz/dfeb53fc8d4ea1d2d972b3b9daee2187

*Layout analyzer output on generated partition layout:* https://gist.github.com/cgtz/608af99f02e30d3ca4702bd16cc36afd

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -45,6 +48,24 @@
 
   private Logger logger = LoggerFactory.getLogger(getClass());
 
+  // Comparators for comparing unallocated capacity in datanodes/disks
```
make this javadoc

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -45,6 +48,24 @@
+  private final Comparator<DataNode> dataNodeCapacityComparator = new Comparator<DataNode>() {
+    @Override
+    public int compare(DataNode o1, DataNode o2) {
+      long o1Capacity = getUnallocatedRawCapacityInBytes(o1);
```
minor: can probably write a function for these three lines in the outer class and reuse it.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +282,195 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+   * Get a list of the nodes in a {@link Datacenter} ordered by amount of free space. Nodes with the same amount
+   * of free space are shuffled.
+   *
+   * @param datacenter
```
describe both param and return

cgtz replied:
done

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/ClusterMapManagerTest.java:
```diff
@@ -149,20 +151,66 @@ public void bestEffortAllocation()
+    ClusterMapManager clusterMapManager = new ClusterMapManager(partitionLayout);
+    List<PartitionId> allocatedPartitions;
+
+    // Allocate a five partitions that fit within cluster's capacity
```
minor: typo, no "a" required.

cgtz replied:
fixed

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/TestUtils.java:
```diff
@@ -345,6 +350,51 @@ public static JSONObject getJsonPartitionLayout(String clusterName, long version
+   *
+   * @param allocatedPartitions the list of partitions to check
+   */
+  public static void checkRackUsage(List<PartitionId> allocatedPartitions) {
```
would this function be useful as a general utility function? What other test might potentially have a use for this function?

cgtz replied:
I will move this to the test class itself.

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/TestUtils.java:
```diff
@@ -345,6 +350,51 @@ public static JSONObject getJsonPartitionLayout(String clusterName, long version
+   * @param allocatedPartitions the list of partitions to check
+   * @param numReplicas how many replicas a partition should have in each datacenter
+   */
+  public static void checkNumReplicasPerDatacenter(List<PartitionId> allocatedPartitions, int numReplicas) {
```
same here.

cgtz replied:
this one too.

cgtz commented:
Addressed comments and added the necessary changes to the PartitionManager CLI.  This can be reviewed more formally now.  Also, we can merge it if it seems good. In testing with layout-analyzer.py, the random selection strategy generated well distributed partition layouts in about a third of the time of the sorted/shuffled list strategy, so I removed code related to the latter strategy.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,149 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+        Disk diskToAdd = Collections.max(diskCandidates, diskCapacityComparator);
+        disksToAllocate.add(diskToAdd);
+        nodesToExclude.add(diskToAdd.getDataNode());
+      }
```
minor: can we do an `else break;`? Is there any point in continuing if one of the iterations doesn't turn up anything?

cgtz replied:
That's a good idea.  If no `diskCandidates` are returned, that means that there are no useable nodes left in the datacenter.

cgtz replied:
That's a good idea, since we would have already checked over all possible candidate nodes in the datacenter.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,149 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
-    ArrayList<Disk> allocatedDisks = new ArrayList<Disk>();
+      Datacenter datacenter, boolean disableRackAwareness) {
+    ArrayList<Disk> disksToAllocate = new ArrayList<Disk>();
+    boolean rackAware = datacenter.isRackAware() && !disableRackAwareness;
```
seems to me like we should throw here if `disableRackAwareness` is `false` but `dataCenter.isRackAware()` is `false`. Otherwise doesn't it mean you are silently no respecting one of the args provided?

cgtz replied:
I was thinking that we should be able to support hardware layouts where some datacenters have rack info and some do not.  Do you think we should throw in this case?

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,149 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+  }
+
+  /**
+   * Return a list of disks for a new partition in the specified {@link Datacenter}.  Does not retry if fewer than
```
should this throw if fewer than `replicaCountPerDatacenter` disks are allocated? Sounds to me like a failure.

vgkholla replied:
Maybe not. Will discuss this with you.

cgtz replied:
Yeah, let's talk about this and other failure cases today.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,149 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
-  // Best effort (or less) allocation of partitions. I.e., size of returned list may be less than numPartitions.
+
+  /**
+   * Perform a best effort (or less) allocation of partitions.
```
looks like this does the best effort only if `bestEffort` is true? That is the right way to do it. Just need to change the documentation.

cgtz replied:
Yes, I'll change the docs.  I forgot to update them.

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/ClusterMapManagerTest.java:
```diff
@@ -137,7 +141,7 @@ public void addNewPartition()
 
   @Test
-  public void bestEffortAllocation()
+  public void bestEffortAllocationNonRackAware()
```
the function name says bestEffort but you pass `false` for the `bestEfffort` arg. I'm a taf confused.

vgkholla replied:
Is it because the variable does not matter when `disableRackAwareness` is `true`?

cgtz replied:
I forgot to change the test name after changing the allocation method's behavior.  I'll look over these test cases and make the names more relevant.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,149 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+        datacenter, disableRackAwareness);
+    if (datacenter.isRackAware() && !disableRackAwareness && (disks.size() < replicaCountPerDatacenter)) {
+      System.err.println("Rack-aware allocation failed for a partition on datacenter:" + datacenter.getName());
+      if (bestEffort) {
```
can you just add this condition in the previous `if`?

cgtz replied:
I will do that.  At first, I wanted to print a message if allocation failed and `bestEffort` was `false`, but that probably isn't needed because we print a similar message anyways on line 381.

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/ClusterMapManagerTest.java:
```diff
@@ -148,21 +152,67 @@ public void bestEffortAllocation()
     assertEquals(allocatedPartitions.size(), 0);
+
+    // Reset hardware layout
+    testHardwareLayout = new TestUtils.TestHardwareLayout("Alpha", true);
```
Any reason to not have this as another test? Resetting seems awkward.

cgtz replied:
I'll move this to another test case.

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/ClusterMapManagerTest.java:
```diff
@@ -148,21 +152,67 @@ public void bestEffortAllocation()
+    allocatedPartitions =
+        clusterMapManager.allocatePartitions(5, replicaCountPerDataCenter + 1, replicaCapacityInBytes, false, true);
+    checkNumReplicasPerDatacenter(allocatedPartitions, 3);
   }
```
Do you have a test for when `disableRackAwareness` is `false` and `bestEffort` is true and there are more replicas than racks? Would that kind test be of use?

cgtz replied:
I will add this test.

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/ClusterMapManagerTest.java:
```diff
@@ -148,21 +152,67 @@ public void bestEffortAllocation()
+  }
+
+  @Test
+  public void bestEffortAllocationRackAware()
```
ok even this says "bestEffort" but the arg `bestEffort` is always `false`. Perhaps there is no relation b/w the function name and the arg but we should rename the function to avoid confusion maybe?

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/ClusterMapManagerTest.java:
```diff
@@ -258,5 +308,50 @@ public void validateSimpleConfig()
+      for (ReplicaId replica : partition.getReplicaIds()) {
+        String datacenter = replica.getDataNodeId().getDatacenterName();
+        Integer replicasInDatacenter = numReplicasMap.containsKey(datacenter)? numReplicasMap.get(datacenter) : 0;
+        numReplicasMap.put(datacenter, replicasInDatacenter+1);
```
Could you apply the formatting? Looks to me like it has not been?

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/TestUtils.java:
```diff
@@ -159,7 +164,7 @@ public static JSONArray getJsonArrayDataNodesRackAware(int dataNodeCount, String
     JSONArray jsonArray = new JSONArray();
     for (int i = 0; i < dataNodeCount; ++i) {
-      jsonArray.put(getJsonDataNode(hostname, basePort + i, sslPort + i, i, hardwareState, disks));
+      jsonArray.put(getJsonDataNode(hostname, basePort + i, sslPort + i, i % 3, hardwareState, disks));
```
does this mean you have hardcoded the number of racks to be three? Any way to make this configurable? Like sending in a `jsonDataCenter` that has the info?

vgkholla replied:
This comment needs to be addressed.

cgtz replied:
This function is used to generate test configurations.  I will add an argument to allow the user to set the number of racks

vgkholla commented on ambry-tools/src/main/java/com.github.ambry.tools/admin/PartitionManager.java:
```diff
@@ -73,6 +79,12 @@ public static void main(String args[]) {
               .withOptionalArg().ofType(String.class);
 
+      String disableRackAwarenessFlag = "disableRackAwareness";
+      parser.accepts(disableRackAwarenessFlag, "Disable rack-aware partition allocation");
```
why not accept this like the other arguments as a `ArgumentAcceptingOptionSpec`?

vgkholla commented:
Finished a pass. Can you provide unit test coverage details? Is every new line covered (including all values of boolean flags)? It should be possible to do so.

Would like to go over the coverage with you.

cgtz commented:
Addressed Gopal's comments in latest commit.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,152 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+   */
+  private List<Disk> allocateDisksForPartition(int replicaCountPerDatacenter, long replicaCapacityInBytes,
+      Datacenter datacenter, boolean disableRackAwareness) {
+    ArrayList<Disk> disksToAllocate = new ArrayList<Disk>();
```
minor: declare this after the illegal arg check.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,152 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+   * @return a list of {@code replicaCountPerDatacenter} or fewer disks that can be allocated for a new partition in
+   *         the specified datacenter
+   */
+  private List<Disk> allocateDisksForPartitionWithRetry(int replicaCountPerDatacenter, long replicaCapacityInBytes,
```
Taking a second look, it seems to me that we need a rethink of these arguments. What is the point of having both `bestEffort` and `disableRackAwareness`?

What if we named `bestEffort` as `retryIfNotRackAware` (or similar) and had only that argument? I think it would make everything simpler. 

For `allocatePartitions()`, if `retryIfNotRackAware` is `false`, we bail out immediately if any of the data centers returns `isRackAware()` as `false` (because it is clearly a lost cause). 
If `retryIfNotRackAware` is `true`, then for the the ones that return `false` for `isRackAware()` we simply disable rack-awareness on the first call to `allocateDisksForPartition()`. For the ones that return `true`, we use the logic in this function.

vgkholla commented:
@cgtz As as aside, always prefer rebase over merge (I see you have a merge commit but it is clean).

cgtz commented:
Updated to use `retryIfNotRackAware` flag.

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,155 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+      Datacenter datacenter, boolean rackAware) {
+    ArrayList<Disk> disksToAllocate = new ArrayList<Disk>();
+    if (!datacenter.isRackAware() && rackAware) {
+      throw new IllegalArgumentException(
```
I don't think this is reachable anymore. This is a private function and the function that calls this one already makes this check.

cgtz replied:
removed

vgkholla commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,155 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+    List<Disk> disks;
+    if (datacenter.isRackAware()) {
+      disks = allocateDisksForPartition(replicaCountPerDatacenter, replicaCapacityInBytes, datacenter, true);
+      if (retryIfNotRackAware && (disks.size() < replicaCountPerDatacenter)) {
```
minor: switch the condition order.

cgtz replied:
switched

vgkholla commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/ClusterMapManagerTest.java:
```diff
@@ -137,7 +149,7 @@ public void addNewPartition()
 
   @Test
-  public void bestEffortAllocation()
+  public void nonRackAwareAllocationTest()
```
why is this called "nonRackAware"?

cgtz replied:
Because the hardware layout does not have rack info in this test case. I'm open to suggestions for better names though.

vgkholla commented:
Minor comments. LGTM

Could you please add the unit test coverage to the PR description? Are all new lines covered?

cgtz commented:
@vgkholla @pnarayanan @nsivabalan Addressed Gopal's comments and added coverage info in description.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,151 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+  /**
+   * Add a set of replicas on a new datacenter for an existing partition.
+   *
+   * @param partitionId The partition to add the new datacenter to
```
"The partition to add to the new datacenter"?

cgtz replied:
will fix

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,151 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+   *
+   * @param partitionId The partition to add the new datacenter to
+   * @param dataCenterName The name of the new datacenter
+   * @param retryIfNotRackAware {@code true} if we should attempt a non rack-aware allocation if a rack-aware one
```
drop the first person subjective: "if a non rack-aware allocation should be attempted..."

cgtz replied:
will fix

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -333,10 +461,9 @@ public void addReplicas(PartitionId partitionId, String dataCenterName) {
-    datacentersToAdd.add(datacenterToAdd);
     List<Disk> disksForReplicas =
-        allocateDisksForPartition(numberOfReplicasPerDatacenter, capacityOfReplicasInBytes, datacentersToAdd);
+        allocateDisksForPartitionWithPotentialRetry(numberOfReplicasPerDatacenter, capacityOfReplicasInBytes,
```
retry makes it sound like trying the same thing (and makes you wonder about the number of times). Perhaps call this method the same as what it was previously, and change the flag to `attemptNonRackAwareOnFailure` or something to that effect?

cgtz replied:
Yeah, I agree that the name is kind of awkward. I'll change it.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,151 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+      long replicaCapacityInBytes, boolean retryIfNotRackAware) {
     ArrayList<PartitionId> partitions = new ArrayList<PartitionId>(numPartitions);
 
     while (checkEnoughUnallocatedRawCapacity(replicaCountPerDatacenter, replicaCapacityInBytes) && numPartitions > 0) {
```
not related to the changes in this patch, but it looks like this check does not take into account the fact that replicas of a partition cannot be on the same node (it simply checks total disk capacity of all nodes in the datacenter with total size required for all replicas).

cgtz replied:
That is true; this is a pretty basic verification.  We do report if we could not allocate the desired number of partitions, but that happens inside of the loop.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,151 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+      int selectionIndex = randomGen.nextInt(selectionBound);
+      DataNode candidate = dataNodes.get(selectionIndex);
+      if (!dataNodesUsed.contains(candidate) && !rackIdsUsed.contains(candidate.getRackId())) {
+        Disk diskWithMostCapacity = getDiskWithMostUnallocatedRawCapacity(candidate, replicaCapacityInBytes);
```
as a general question - what is the rationale in choosing the disk with the most free space really? Basically choosing worst fit allocation over best fit allocation? I understand that this way one might allocate the least loaded disk (assuming all disks have even capacities to begin with). Setting that aside, is worst fit a better allocation strategy?

cgtz replied:
I did not think too much about this; the original algorithm also chose the disk with the most free space. However, choosing the disk with the most free space will help to evenly distribute partitions across all the disks in a single machine.  Do you think this would improve throughput at all compared to trying to use the minimum number of disks per machine?

pnarayanan replied:
Was not thinking in terms of even distribution - but in terms of future allocations. Best fit is more intuitive at first, basically going with the disk with the least free space that can accommodate the partition, so that future allocations will have a better chance of finding space. I think this matters only if partitions could be of different sizes. For example, if we wanted to allocate a 100G partition, and the algorithm finds two disks with 100G and 400G respectively, then the 400G disk will be chosen using the current scheme. However, if the next partition to allocate is of size 400G, then there wouldn't be any candidates available because of the worst fit allocation strategy.

However, it is recommended to use the same size for all partitions (at least within a scaling unit), and so worst fit would probably be as good as best fit, with the added benefit of load balancing.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,151 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+      if (diskCandidates.size() > 0) {
+        Disk diskToAdd = Collections.max(diskCandidates, diskCapacityComparator);
+        disksToAllocate.add(diskToAdd);
+        nodesToExclude.add(diskToAdd.getDataNode());
```
A subtle difference from the referenced algorithm is that in our use case, the events are *dependent* - that is here we *exclude* the candidate (which we have to, because no more than one replica can be on a node). In the referenced example, the chosen candidates *remain* in the candidate list - the chosen bin for a ball is eligible to be chosen for subsequent balls. I'd bet that the dependence makes some difference as far as the point of the article is concerned, but perhaps not too much.

cgtz replied:
Yes, the optimality proof in the article probably isn't valid when nodes are excluded.  Empirically however, it seems like the allocations generated are well balanced.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -261,45 +283,151 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+      List<Disk> diskCandidates =
+          getRandomDiskCandidateSample(dataNodes, nodesToExclude, replicaCapacityInBytes, rackAware, NUM_CHOICES);
+      if (diskCandidates.size() > 0) {
+        Disk diskToAdd = Collections.max(diskCandidates, diskCapacityComparator);
```
It looks like `getRandomDiskCandidateSample` could encapsulate this logic and simply return the chosen disk. Did you have it this way because previously two algorithms worked on the sample? If so, I think encapsulating all this in one method will simplify things (no need for that method to maintain or return a list). Perhaps that method could be named differently as well.

cgtz replied:
Yes, that's a good idea.  I'll change the method to just return the best disk.

cgtz commented:
Addressed @pnarayanan 's comments.  I also fixed an issue where I was using the free space on the disks in the random sample to choose instead of the free space on the entire node. All of the commits are showing up right here because I just rebased on top of master.

pnarayanan commented on ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java:
```diff
@@ -334,18 +339,17 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
-        Disk diskToAdd = Collections.max(diskCandidates, diskCapacityComparator);
-        disksToAllocate.add(diskToAdd);
-        nodesToExclude.add(diskToAdd.getDataNode());
+      Disk diskCandidate =
```
at this point this is really `bestDisk` - is it not?

cgtz replied:
Yeah, let me fix that right now.

pnarayanan commented:
Minor comment, LGTM otherwise.

cgtz commented:
@pnarayanan @vgkholla Updated PR with minor changes.

vgkholla commented:
Minor comment. LGTM

Have you tried using the CLI tool and made sure all the options you introduced work as expected? Have you done a test allocation and examined the output? It might be useful to do so and attach a gist if possible.

cgtz commented on ambry-clustermap/src/test/java/com.github.ambry.clustermap/ClusterMapManagerTest.java:
```diff
@@ -148,24 +161,96 @@ public void bestEffortAllocation()
+          + "and retryIfNotRackAware is false");
+    } catch (IllegalArgumentException e) {
+      // This should be thrown
+    } catch (Throwable e) {
```
fixed

cgtz commented:
Yes, I have tried out the tool from the command line and done a test allocation. I will post gists of relevant output in the original PR comment

pnarayanan commented:
LGTM. Merging.
